File: fs/symlink-dependency/src/safeJoinModulesDir.ts
import path from 'node:path'

import validateNpmPackageName from 'validate-npm-package-name'

// Joins `modulesDir` with a dependency alias and guarantees the result
// stays a direct child of `modulesDir`. The alias becomes a directory
// name inside `node_modules`, so it must be a valid npm package name: a
// single `name` or `@scope/name` of URL-friendly characters with no
// leading `.` or `_`, and not a reserved name. That rejects
// path-traversal (`../x`), absolute, and pnpm-owned aliases (`.bin`,
// `.pnpm`, `node_modules`) before they can escape `modulesDir` or
// overwrite pnpm's own layout. The containment check is kept as a
// belt-and-suspenders guard for any platform-specific join behavior the
// name check might not anticipate.
//
// Earlier passes reject such aliases at manifest-read and resolution
// time, but this layer also runs for paths reconstructed from lockfiles
// and snapshots, so the check stays here as a final guarantee.
export function safeJoinModulesDir (modulesDir: string, alias: string): string {
  if (!validateNpmPackageName(alias).validForOldPackages) {
    throw invalidDependencyNameError(modulesDir, alias)
  }
  const link = path.join(modulesDir, alias)
  const resolvedDir = path.resolve(modulesDir)
  const resolvedLink = path.resolve(link)
  if (resolvedLink === resolvedDir || !resolvedLink.startsWith(resolvedDir + path.sep)) {
    throw invalidDependencyNameError(modulesDir, alias, resolvedLink)
  }
  return link
}

function invalidDependencyNameError (modulesDir: string, alias: string, resolvedLink?: string): Error & { code: string } {
  const detail = resolvedLink ? ` (it resolves to ${resolvedLink})` : ''
  const error = new Error(`Refusing to place a dependency under ${modulesDir} with the invalid alias ${JSON.stringify(alias)}${detail}`) as Error & { code: string }
  error.code = 'ERR_PNPM_INVALID_DEPENDENCY_NAME'
  return error
}

File: pacquet/crates/package-manager/src/safe_join_modules_dir.rs
//! Join a `node_modules` directory with a dependency alias and reject
//! aliases that aren't valid npm package names before the join, because
//! the alias becomes a directory name inside `node_modules`. Mirrors
//! pnpm's
//! [`safeJoinModulesDir`](https://github.com/pnpm/pnpm/blob/main/fs/symlink-dependency/src/safeJoinModulesDir.ts)
//! and routes through the same [`is_valid_dependency_alias`]
//! check pacquet applies to direct-dependency aliases at resolution
//! time, so the hoisted restore path enforces the boundary the
//! resolution path already enforces.

use derive_more::{Display, Error};
use miette::Diagnostic;
use pacquet_resolving_deps_resolver::is_valid_dependency_alias;
use std::path::{Path, PathBuf};

/// A dependency alias that would escape `modules` or collide with
/// pnpm's own `node_modules` layout. Surfaces pnpm's
/// `ERR_PNPM_INVALID_DEPENDENCY_NAME`.
#[derive(Debug, Display, Error, Diagnostic)]
#[display("Refusing to place a dependency under {} with the invalid alias {alias:?}", modules.display())]
#[diagnostic(code(INVALID_DEPENDENCY_NAME))]
pub struct InvalidDependencyAliasError {
    pub modules: PathBuf,
    #[error(not(source))]
    pub alias: String,
}

/// `modules.join(alias)` guarded by a package-name validity check.
/// Returns [`InvalidDependencyAliasError`] when `alias` is not a valid
/// npm package name.
pub fn safe_join_modules_dir(
    modules: &Path,
    alias: &str,
) -> Result<PathBuf, InvalidDependencyAliasError> {
    if !is_valid_dependency_alias(alias) {
        return Err(InvalidDependencyAliasError {
            modules: modules.to_path_buf(),
            alias: alias.to_owned(),
        });
    }
    Ok(modules.join(alias))
}

#[cfg(test)]
mod tests;
